### ISSUE TITLE
Eventually implement support for the PEP 735 "dependency groups"

### ISSUE THREAD

**Issue by webknjaz** [jazzband/pip-tools#2062]
#### What's the problem this feature will solve?

It's just a new source of deps and an extension to the first round of the `pyproject.toml` runtime deps resolution.

#### Describe the solution you'd like

$sbj. If somebody's interested in making a PoC, it's probably fine to start drafting something already. Need to wait for the PEP acceptance to actually merge it in, though.

#### Alternative Solutions

... not doing this?

#### Additional context

Refs:
* https://peps.python.org/pep-0735/
* https://discuss.python.org/t/pep-735-dependency-groups-in-pyproject-toml/39233/237
* POC for pip: https://github.com/pypa/pip/compare/main...sirosen:pip:pep-735-support

**Comment by chrysle** [comment]
What are you planning to have for dependency group support? An option to collect only such dependencies, or alongside with normal and build deps? And the possibility to collect only specified dependency groups?

**Comment by webknjaz** [comment]
Same as #1960 attempted, just in the proper place. Semantically, dependency groups don't imply including runtime dependencies so they should definitely be collected and output separately. However, we shouldn't actively prevent combining them with other deps (runtime, or runtime+extra, or other dependency groups; though the latter seems to be covered by the PEP itself).

**Comment by chrysle** [comment]
> However, we shouldn't actively prevent combining them with other deps

IIUC that _would_ mean actually creating an option like `--include-dependency-groups`/`--all-dependency-groups`, next to `--only-dependency-groups` and `--dependency-group GROUP`?

**Comment by webknjaz** [comment]
Aren't the past two the same?

**Comment by chrysle** [comment]
> Aren't the past two the same?

Yes, this was unintended. Is the rest of my assumption valid?

**Comment by webknjaz** [comment]
I think so. It of course could be iterated on during a PoC review..

**Comment by webknjaz** [comment]
UPD: said PEP is accepted as of two weeks ago. So we should implement support for it.

**Comment by davidism** [comment]
Is there an idea for how this will work? I'd expect to be able to move each `.in` template into a separate dependency group instead. Then `pip-compile pyproject.toml --group tests` would read the `tests` group from `pyproject.toml` and output to `tests.txt` (or to `-o {name}`). Does that sound reasonable?

Currently I have different environments as separate templates: `requirements/test.in`, `requirements/typing.in`, etc. Now I'd have `tests` and `typing` dependency groups instead but otherwise get the same behavior.

**Comment by webknjaz** [comment]
> Is there an idea for how this will work? I'd expect to be able to move each `.in` template into a separate dependency group instead.

Yeah, that's kinda the use-case this PEP solves.

> Then `pip-compile pyproject.toml --group tests` would read the `tests` group from `pyproject.toml` and output to `tests.txt` (or to `-o {name}`). Does that sound reasonable?

This needs to be defined, I think. While working on support for the distribution package build dependencies (#1681 / https://pip-tools.readthedocs.io/en/stable/#maximizing-reproducibility), we realized that there would be cases where one would want to extract those dependencies in isolation and cases of getting combined deps (`--build-deps-for=wheel` / `--all-build-deps` / `--only-build-deps `).

There's a precedent that doing `pip-compile pyproject.toml requirements.in` works and sources the deps from several places. That can also be combined with `--extras`, IIRC.

So to proceed meaningfully, we need to define the semantics and UX for all of these combinations. Do we need `--[no-]runtime` or some other ways to distinguish what the users would like to encode in their constraint file generation requests? Additionally, I would like to keep #2124 in mind as well, while answering these questions.

> Currently I have different environments as separate templates: `requirements/test.in`, `requirements/typing.in`, etc. Now I'd have `tests` and `typing` dependency groups instead but otherwise get the same behavior.

Simlarly, I have even more envs defined, mostly one per tox target with multiple constraint files (aka DIY lock files) for each, since combinations of different platforms/runtimes can count as separate environments (fun stuff: https://github.com/ansible/awx-plugins/tree/c7fc0a1/dependencies/lock-files#readme).

**Comment by merwok** [comment]
What is the meaning of `$sbj.` ?

**Comment by webknjaz** [comment]
@merwok sbj == subject. As in a reference to what's written in the issue title.

**Comment by webknjaz** [comment]
@sirosen I think we have a tradition of double colons. I.e. `::`. So it only seems natural to use that as a separator.

**Comment by sirosen** [comment]
Okay. I don't like that it will drift from `pip` syntax but if it's consistent within `pip-tools` that might be best.

Just to lay out how it would look, here are the single vs double colon options, plus a third idea:

### Single Colon
- `pip install --group './foo/pyproject.toml:bar'`
- `pip-compile --group './foo/pyproject.toml:bar'`

### Double Colon
- `pip install --group './foo/pyproject.toml:bar'`
- `pip-compile --group './foo/pyproject.toml::bar'`

### 1-2 Colons

i.e. Make `::` and `:` both work as aliases/alternatives.

- `pip install --group './foo/pyproject.toml:bar'`
- `pip-compile --group './foo/pyproject.toml:bar'`
- `pip-compile --group './foo/pyproject.toml::bar'`

---

I'll work on an PR as soon as I'm able, using the double-colon syntax. If the 1-2 colon option looks nice, I'd be happy to pursue that as well.

**Comment by sirosen** [comment]
I'm not sure if I was the addressee for that comment, but I'm not sure what the benefit of `::` is in this context.

I was planning to lay low on this one until I get some time to try putting the PR together, and then advocate for the "1-2 colons" version but be ready to pivot to something else if `pip-tools` maintainers have strong opinions.

**Comment by Gankra** [comment]
Another update from `uv pip compile`:

We're planning to move forward with just support for `:` but if you ship `::` we'll add support both so that we're consistent with y'all and pip.

---
 
We're also planning to add subtle group context in `# via` comments (don't have our hearts set on the syntax here but that's what we have for now). So an invocation like this:

```
uv pip compile pyproject.toml --group subdir/pyproject.toml:foo
```
produces the following (note the `:foo` after `subdir/pyproject.toml`)

```
# This file was autogenerated by uv via the following command:
#    uv pip compile pyproject.toml --group subdir/pyproject.toml:foo
iniconfig==2.0.0
    # via mysubproject (subdir/pyproject.toml:foo)
typing-extensions==4.10.0
    # via myproject (pyproject.toml)
```

**Comment by webknjaz** [comment]
@Gankra to be in line with the build system extraction refs, the output would need to be
```diff
 # This file was autogenerated by uv via the following command:
 #    uv pip compile pyproject.toml --group subdir/pyproject.toml:foo
 iniconfig==2.0.0
-    # via mysubproject (subdir/pyproject.toml:foo)
+    # via mysubproject (subdir/pyproject.toml::dependency-groups.foo)
 typing-extensions==4.10.0
     # via myproject (pyproject.toml)
```

(see https://pip-tools.readthedocs.io/en/stable/#maximizing-reproducibility for the build deps refs output example that has things like `via my-cool-django-app (pyproject.toml::build-system.requires)`)

**Comment by webknjaz** [comment]
> Is there a particularly compelling reason not to support the upstream syntax here? (supporting two colons as well seems fine to me, but I don't have as much context as you)

So my reasoning was that (1) pip-tools doesn't really have an “upstream”. It happens to rely on parts of pip, but it performs semantically different tasks and the `pip-complile` CLI mostly doesn't seem to mimic pip, short of the `--pip-args` hack.
Another reason (2) is that we have `::` in the output.

In general, I don't think I'm strongly against a single colon. But then again, there are a few other patterns existing in the CLI that relate to extracting things out of `pyproject.toml` (like Poetry-specific sections, regular PEP 621, build deps) and they mostly allow a single `pyproject.toml`, not several IIRC, with various targets being requested via CLI options that don't include the file path.
I feel like following `pip` might cause UX confusion in this area. But at the same time, I don't have a better suggestion.

**Comment by webknjaz** [comment]
@sirosen I realized that you mentioned planning to send a PR so I assigned you.

**Comment by Gankra** [comment]
> [@Gankra](https://github.com/Gankra) to be in line with the build system extraction refs, the output would need to be
> 
> ```
>  # This file was autogenerated by uv via the following command:
>  #    uv pip compile pyproject.toml --group subdir/pyproject.toml:foo
>  iniconfig==2.0.0
> -    # via mysubproject (subdir/pyproject.toml:foo)
> +    # via mysubproject (subdir/pyproject.toml::dependency-groups.foo)
>  typing-extensions==4.10.0
>      # via myproject (pyproject.toml)
> ```
> 
> (see https://pip-tools.readthedocs.io/en/stable/#maximizing-reproducibility for the build deps refs output example that has things like `via my-cool-django-app (pyproject.toml::build-system.requires)`)

Oh right, a bit of pushback on this: while this argument is reasonable, I think there's a key difference here. `subdir/pyproject.toml:foo` reflects the actual syntax the user can pass on the pip/uv-pip/pip-compile CLI to activate the group. So the prettier syntax actually is more useful to use here.

There's no ability to pass `build-system.requires` on the CLI, so it has no such benefit/constraint.

**Comment by reactive-firewall** [comment]
> UPD: said PEP is accepted as of two weeks ago. So we should implement support for it.

@webknjaz is there any update on releasing the PR from @sirosen (e.g., #2175) now that [PEP-735](https://peps.python.org/pep-0735/) was finalized?

Additional refs:
* https://packaging.python.org/en/latest/specifications/dependency-groups/#dependency-groups
* https://pip.pypa.io/en/stable/user_guide/#dependency-groups (official `pip` support)

**Comment by sirosen** [comment]
I've become a pip-tools maintainer in the intervening year. My focus has been primarily on fixes and paying down tech debt -- new features like this have been more or less on hold.

I think it's likely to get included in one of the next few releases. We'll have to settle on exactly what we want in the annotation comments ("via"), but I think that's the only outstanding issue to address.

**Comment by webknjaz** [comment]
@sirosen I think it's probably fine to support both variants as inputs. I was mostly worried about the output. Although, I can see how having two ways to do the same might end up being confusing.

**Comment by DarkSharkTornado** [comment]
## Proposed Implementation Plan for PEP 735 Dependency Groups

I'd like to contribute an implementation for PEP 735 (dependency-groups) support in pip-tools.

### Scope
- Parse `[dependency-groups]` sections from `pyproject.toml`
- Add CLI flags: `--group <name>` / `--all-groups`
- Propagate group resolution to `pip compile` output
- Handle transitive dependencies within groups
- Add comprehensive tests matching existing test patterns

### Approach
1. Extend `pip_tools/resolver.py` to recognize dependency-groups
2. Update `pip_tools/cli.py` with new flags
3. Mirror the resolution logic already used for `--extra` and `--constraint`
4. Follow the existing test structure in `tests/`

### Timeline
- Week 1: Core parsing + resolution logic
- Week 2: CLI integration + edge case handling
- Week 3: Tests, docs, and review

Happy to discuss scope and adjust before starting. I've contributed to Python packaging tools before and am comfortable with the pip codebase.

Looking forward to contributing!

**Comment by webknjaz** [comment]
@DarkSharkTornado since your account was created just 19 minutes ago, I'll just leave https://dontquotetheai.com and https://pip-tools.rtfd.io/en/stable/contributing/#llm-generated-contributions for you to learn. This is not the best way to get started with your first open source contribution. Be authentic and genuine, start small, find your motivation and be ready to explain yourself without using your LLM or an agent.
